### ISSUE TITLE
Making it easier to add new iteration spaces

### ISSUE THREAD

**Issue by hiker** [stfc/PSyclone#109]
Hi,

for my application (which I am psyclonifying atm) I need to add at least two new iteration spaces (internal_pts + halo N(orth)/S(outh); and internal_pts + halo W/E  - maybe later I also need extnernal_pts_NS and similar WE, or perhaps only the latter ... not sure atm).

At this stage this is rather inconvenient to add an iteration spcae: the above iteration spaces appear to be rather limited in use, and in order to pass on my work to others I would need to add the iteration spaces to dl_esm_if, then to psyclone. The latter would involve creating up to 5 version of it (depending on offset I guess - maybe only 1 case if I can use offset_any??). And if I need to have the code to work with an official psyclone release (as opposed to a copy of my fork), I then also need to add test cases, review process, ... And all I am doing is adding rather application specific code (to support the iteration space) to psyclone

Would you consider it useful if psyclone could be extended locally? I.e. adding some python files to the solution (iteration_space.py?), which is then picked up by psyclone, and allows new iteration spaces locally to an application? It doesn't solve the dl_esm_if synchronisation problem, but at least removes some very application specific code from psyclone.

I did not really thought about implementation (not even if this is even feasible), just thought to discuss this first - and I am happy to suggest a more detailed approach if you are interested in this idea.

Cheers,
   Joerg

**Comment by rupertford** [comment]
Hi Joerg,

Are these really different iteration spaces, or are these the same iteration space with different types of redundant computation (depending on the sort of stencil access you are expecting)?

Forgive me if I am mistaken but I would say that you should never need to write into the halo in order for your code to be correct, it is the job of halo exchanges to ensure that halos are consistent when they are read. For performance reasons you might want to write redundantly into the halo, but that is an optimisation issue.

Regards

Rupert

**Comment by hiker** [comment]
Ah, wrong word from my side ... kind of:
the algorithm updates the boundary (which is stored in the halo, which is why I used the wrong word). I have two different patterns there:
```
do j=2,ny-1
  do i=1, nx
    call swlon_update(i, j,...)
  enddo
enddo
do  i=2,nx-1
  do j=1,ny
    call swlat_update(i, j, ...)
  enddo
enddo
```
Yes, I am aware of the 'wrong' ordering in the 2nd loop :)  So the algorithms does longitudinal updates including the West- and East-boundary fields (stored in the halo), then latitudinal updates including the North- and South-boundaries. 

Some of the kernel will explicitly test if the (i,j) point is on the boundary and handle those points differently.

It might well be that after more restructuring of the code I might only need external_pts as iteration space, but I am not entirely sure. For now I start with converting the code one kernel at a time, to be able to have bitwise identical results (which is great for testing), so I kind of need those strange iteration spaces. I just added an internal_we_halo iteration space to dl_esm_inf and psyclone, and my application is working fine now, so I am making great progress ;)

Once I have all converted the main maths to kernels, I'll look at optimisation, and that obviously will include getting if statements out of the kernels. I suspect that I will need one iteration space 'NS' boundary only, and one 'WE' boundary only (again to avoid if statements in the kernels), but I need more time to evaluate this (or now that I think about it - maybe even 4 spaces: N, S, W, E boundary updates). Sorry, I need more time to study the code to decide what will be best.

But I think the main point of this ticket: I don't think that psyclone should have all the iteration spaces hard coded. I am not sure at this stage (which is why I opened this ticket) where this should go ... maybe be part of the runtime library? For example:

if you use the dl_esm_if runtime library, this library should somehow define all the iteration spaces as data. This would then be queried by psyclone to create the correct loops. As a quickly made up example, dl_esm_if could contain:
```
it_spaces=["all_pts", "internal_pts", "external_pts"]

getBoundaries(iteration_space):
  # Returns the outer and inner loop indices
  if iteration_space=="all_pts":
    return ( ("0", "+1"), ("0", "+1") )   # --> j=0, ny+1, i=0, nx+1,
  elif iteration_space=="internal_pts":
   return ( ("1", ""), ("1", "") )    # --> j=1, ny, i=1, nx
```
Psyclone could get the list of valid iteration spaces from a python module of the runtime library, validate any user code against this, and also query the loop boundaries from this runtime-specific python module. Considering the 4 different field types (and potentially 4 different offset types) - I need some proper design, but I think all of this could go into an runtime specific python module instead of into psyclone. This would decouple the iteration space a lot from the psyclone internals.

Not sure if this makes sense or even is feasible :)

**Comment by rupertford** [comment]
Apologies for misunderstanding what you were saying, I'm working on redundant computation support at the moment so that was in my brain.

I think I understand your point. I think you want to be able to define a stencil access in a simple modular way and ensure that it only needs to be specified once, rather than separate in the infrastructure and in PSyclone. I like this idea in three ways.

1) modularization/decoupling is good. If we can isolate the sort of things people will like to add/modify in one place that would make development much easier. I think our modularization is very hit and miss at the moment as we've not really been thinking about how to make it easy to extend existing functionality.

2) defining stencils in a simple way makes it clearer what sort of stencil is being used.

3) linking information between the infrastructure and psyclone helps with correctness and makes it easier to update. We talked about this a little with the Met Office but did not get very far, so at the moment the Met Office write their infrastructure and we keep a separate copy of the concepts and names in PSyclone.

So yes, some form of configuration module, or modules sound very useful.

**Comment by hiker** [comment]
I started to have a look at this. I created a per-kernel 'KernelIterationSpaceInfo' class that would store the information about a kernel and which would be part of the runtime library. This would define and handle information like iteration space, and other information (e.g. in the case of gocean the offset type and field_types), and hide all of this from PSyclone. The current gocean1p0 file would mostly become a thin wrapper around this information class. For example, the kernel information:
```
integer :: ITERATES_OVER = INTERNAL_WE_HALO
integer :: index_offset = OFFSET_NE
```
would all be handled by function(s) in the KernelIterationSpaceInfo, NOT by PSyclone at all. 

This would allow the runtime library to change or add iteration spaces, and other information (like index_offset), without affecting PSyclone.

We would need (at least) two classes: one instance of a class to store information about all kernels in one invoke call (to e.g. check that all kernels use the same index_offset), and one instance of a class per kernel that stores its iteration space, field types etc. PSyclone would only find the right node for the meta_args, and pass this to the right runtime class.

The KernelIterationSpaceInfo class would have functions that PSyclone can query to find out how many nested loops to create (maybe also which variable names to use), and their boundary etc. This allows the runtime library to add for example 1d iteration spaces (like mentioned above I would need four 1-d iteration spaces for each boundary), and it would also make it easier to e.g. create a 3d runtime library.

Now here the big questions:
1) Is this approach generic enough? I remember Andrew creating a 3d-branch in dl_esm_inf ;) So would this still be useful for this? Not to mention the dynamo* API. No idea if it even makes sense to aim at only having one generic PSyclone class that could create code for all those different APIs.
2) The even more important question: I am not yet fully understanding all the classes used in gocean1p0, but my feeling is that 90% of the code from gocen1p0 will be moved into a runtime-specific class (many lines are just handling the various loop boundaries depending on field type, iteration space and offset type). This begs the question: does it actually make sense to create this runtime-specific information classes, should we instead perhaps just move the whole gocean1p0 file into the runtime library, and add some 'infrastructure' to tell PSyclone where to find this (e.g.:
```
export PSYCLONE_API=/PATH/gocean1p0:/OTHER_PATH/gocean0p1:/THIRD_PATH/dynamop1
psyclone ...
```
PSYCLONE_API is a comma-separated class pointing to the various runtime libraries. PSyclone would iterate through all paths to find config files which would specify the API name, and which classes to instantiate depending on API (i.e. recreating the various factories that are in PSyclone).

Now on the other hand, this might wreak havoc with testing - since e.g. all gocean1p0 tests would also need to be part of the gocean runtime library, i.e. outside of PSyclone.

In short - I need someone with a much better and deeper understanding of PSyclone to provide some guidance ;) Should I keep on working on my approach? Or look at moving the API specific files to external directories? Or abandon the whole idea, and accepting the fact that we there is this odd dependency between runtime library and PSyclone?

**Comment by hiker** [comment]
I am closing this ticket, and will do a series of smaller tickets to improve the situation step by step ... starting with #131 and the corresponding pull request #180; ticket #129.
